### ISSUE TITLE
logstash dies with OpenBreaker exception

### ISSUE THREAD

**Issue by webmstr** [logstash-plugins/logstash-input-lumberjack#36]
About twice a day, logstash 1.5.2 will stop processing requests.  It logs this message:

    {:timestamp=>"2015-07-07T09:25:07.113000-0700", :message=>"Exception in lumberjack input thread", :exception=>#<LogStash::CircuitBreaker::OpenBreaker: for Lumberjack input>, :level=>:error}

The only reference I see for this is [in the code](https://github.com/logstash-plugins/logstash-input-lumberjack/blob/master/lib/logstash/inputs/lumberjack.rb#L97), where it says, "it's safe to ignore".  Either that's not true, or not implemented well.

The comment also says that it's a result of too many errors.  No other errors are logged before this message appears.

Restarting logstash by itself doesn't help much, because it doesn't seem to like waking up with 500 shippers trying to connect.  It will say "Lumberjack input: the pipeline is blocked, temporary refusing new connection" and throw more of these OpenBreaker errors (though not always fatally?)  Sometimes I have to shut down all the shippers and bring them back up slowly, allowing the backlog to clear before bringing more shippers online.

**Comment by webmstr** [comment]
I'm now running a cron job to restart logstash every three hours, because if it's dead when I get to the office, it won't catch up.

**Comment by driskell** [comment]
> The comment also says that it's a result of too many errors. No other errors are logged before this message appears.

I think its badly phrased. The history here is that if events were coming in faster than Logstash could handle, lumberjack connections would begin to block too long trying to push events into the Logstash queue. If Logstash forwarder timed out while its connection was blocked, it would disconnect, reconnect, and send events again, and potentially block again. Problem is a blocked connection never frees. So if the rate at which forwarders begin reconnecting exceeds the rate at which connections are becoming unblocked and becoming freed, the connection count would increase to the point where Logstash would crash (`Too many open files` or `OutOfMemory`).

To prevent the crash, latest lumberjack now times out connections if they can't push onto the Logstash queue. However, clients would still keep reconnecting and resending events and potentially duplicating events. First try only manages to push 5 of 50 events, second try managed 10 of 50, third try managed 2 of 50 - but its the same 50 every time. Basically all events in the spool must be pushed within the timeout or they are ALL retried.

To prevent that storm of event duplication if Logstash is going to slow - the breaker was implemented to throw up a barrier until Logstash caught up.

TL;DR:
At the time the breaker throws, your event incoming rate is bigger than Logstash handle. You likely need more resources.
However, if it's just at one point in day where rate is too big, and you don't mind a slight delay while it catches up, [Log Courier](https://github.com/driskell/log-courier) might help.

**Comment by webmstr** [comment]
I added a second logstash server to the mix yesterday, which I don't think has helped.  I had already increased my timeout to 60 seconds and decreased the batch size to 100 events.

logstash-forwarder just stops - no retry messages are seen as I've seen before (with debug enabled).  Since it doesn't failover to the other logstash server, now I have to restart all 500 shippers via cron.

**Comment by webmstr** [comment]
Ah, crap.  LS_WORKER_THREADS was set high, but the init script didn't use it.  Sigh.

**Comment by webmstr** [comment]
Except I don't see that any of the packaged init script handle "-w" or "--filterworkers"...

https://github.com/elastic/logstash/tree/02886972192c3d9972f867c80d069b1bd719a917/pkg

**Comment by driskell** [comment]
You put it in LS_OPTS in the /etc/default or /etc/sysconfig file for logstash.

**Comment by driskell** [comment]
```
LS_OPTS="-w 10"
```

**Comment by udaivizury** [comment]
The problem continues to haunt. Is there any solution available so far?
Thanks a lot.
Udai

**Comment by yuz11** [comment]
Is it default config of the filterworker thread is 1?should set to a higher num by -w num phase while starting the process. I met with this problem today,once I setup more LSF process, then the LS get jam after a while.Now I use -w 10 to start process seems to work. It's able to avoid the jam for now.should wait and see what's going to happen next.BTW,should get the jvm of logstash enough heap mem

**Comment by shashidharrao** [comment]
Could you please let me know how do i pass the LS_OPTS="-w 10" in the zip version, i am not using the rpm packages, what i tried added the OPTS in the logstash startup script but still i am facing the same issue 

{::timestamp=>"2015-08-29T18:01:34.538000-0700", :message=>"Lumberjack input: the pipeline is blocked, temporary refusing new connection.", :level=>:warn}.

thank you.

**Comment by fizerkhan** [comment]
+1
